### ISSUE TITLE
Cannot create output from generated model - tries loading wrong variables

### ISSUE THREAD

**Issue by cbonitz** [ibab/tensorflow-wavenet#85]
I am trying to generate.py using `python generate.py --samples 16000 ./model.ckpt-3999`

However, the loading process fails with the following error:
```
Caused by op u'save/restore_slice', defined at:
  File "generate.py", line 174, in <module>
    main()
  File "generate.py", line 113, in main
    saver = tf.train.Saver(variables_to_restore)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/saver.py", line 986, in __init__
    self.build()
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/saver.py", line 1015, in build
    restore_sequentially=self._restore_sequentially)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/saver.py", line 620, in build
    restore_sequentially, reshape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/saver.py", line 357, in _AddRestoreOps
    tensors = self.restore_op(filename_tensor, saveable, preferred_shard)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/training/saver.py", line 270, in restore_op
    preferred_shard=preferred_shard))
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/io_ops.py", line 204, in _restore_slice
    preferred_shard, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_io_ops.py", line 359, in _restore_slice
    preferred_shard=preferred_shard, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 749, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2380, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1298, in __init__
    self._traceback = _extract_stack()

NotFoundError (see above for traceback): Tensor name "wavenet/causal_layer/Variable" not found in checkpoint files ./model.ckpt-3999
	 [[Node: save/restore_slice = RestoreSlice[dt=DT_FLOAT, preferred_shard=-1, _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_save/Const_0, save/restore_slice/tensor_name, save/restore_slice/shape_and_slice)]]
	 [[Node: save/restore_slice_48/_117 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/gpu:0", send_device="/job:localhost/replica:0/task:0/cpu:0", send_device_incarnation=1, tensor_name="edge_36_save/restore_slice_48", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
```
The tensors in my checkpoint file have names that look deliberate, for example, `//wavenet/causal_layer/filter/Adam`

However, some println debugging shows that the saver in generate.py is trying to load very generic variable names like this:
```
wavenet/causal_layer/Variable
wavenet/dilated_stack/layer0/Variable
wavenet/dilated_stack/layer0/Variable_1
wavenet/dilated_stack/layer0/Variable_2
wavenet/dilated_stack/layer0/skip
wavenet/dilated_stack/layer1/Variable
```

I'm using tensorflow master compiled against Cuda 8.0 RC.

Not sure if this is an issue with my setup or a bug. Any help is greatly appreciated.

**Comment by ibab** [comment]
Are you sure that the `wavenet_params.json` file wasn't changed between training and generation?
The fact that it's failing on layer 16 and not earlier makes me think that.
You could try training with the current `wavenet_params.json`, and then generating with the new checkpoint.

**Comment by cbonitz** [comment]
Thanks for your reply. I'm sorry for misleading you, I accidentally posted a stacktrace from a session where I tried filtering out the first variable that failed - which was `wavenet/causal_layer/Variable`. I now posted the correct stacktrace.
The JSON file was not changed.

**Comment by maxhodak** [comment]
I'm seeing this too. The `wavenet_params.json` definitely hasn't changed.

**Comment by mixcoder** [comment]
hi @cbonitz  , I encounter the same problem, have you fixed it now?

**Comment by lmaxwell** [comment]
the same error too

**Comment by MikalaiDrabovich** [comment]
same error, starnge but it is not appearing with a model generated 3 days ago

**Comment by shin1985** [comment]
same error, the wavenet_params.json definitely hasn't changed.

**Comment by gorinars** [comment]
same error

**Comment by ibab** [comment]
It seems that the generator isn't assigning the right names to each of the variables during construction.
But it seems to work when I test it with the newest version.
I'll have a closer look.

**Comment by ibab** [comment]
Sorry, I was using the version of the project in #88.
When I use the newest `master`, I can reproduce it.
I've taken a look at `wavenet.py` and found the problem.
Some of the variables have been created like this:
```python
weights_dense = tf.Variable(
    tf.truncated_normal(
        [1, dilation_channels, in_channels], stddev=0.2, name="dense"))
```
Here, the name has been given to the initializer op instead of the variable.
Instead, the name has to be part of the variable definition:
```python
weights_dense = tf.Variable(
    tf.truncated_normal([1, dilation_channels, in_channels], stddev=0.2),
    name="dense")
```
I've completely rewritten the variable handling in #88 such that these types of mistakes are going to be impossible in the future.
Once we've merged #88 this problem will be gone.

**Comment by cbonitz** [comment]
Sorry again for the confusion. Thanks a lot for tracking down the root cause. I can confirm that it works for me on branch restructure (#88).
